File: go/native/extension.go
package native

import (
	"encoding/json"
	"fmt"

	"github.com/egoist/quickgui/go/host"
)

var extensionListeners = map[uint32]func(string){}

// ExtensionSession is an application-owned native service. Its callbacks run on
// the UI goroutine. Close it on that goroutine when the service is no longer used.
type ExtensionSession struct {
	name   string
	id     uint32
	closed bool
	ready  bool
}

// OpenExtension starts an imported service extension asynchronously. The provider
// retains its event stream until Close; command replies never remove the stream.
func OpenExtension(name string, options any, changed func(string), ready func(error)) *ExtensionSession {
	s := &ExtensionSession{name: name}
	fail := func(err error) {
		wasClosed := s.closed
		s.closed = true
		if !wasClosed && ready != nil {
			ready(err)
		}
	}
	if !appReady {
		fail(fmt.Errorf("call native.Run before starting an extension"))
		return s
	}
	encoded, err := json.Marshal(options)
	if err != nil {
		fail(err)
		return s
	}
	s.id = allocateRequest()
	extensionListeners[s.id] = changed
	pendingReplies[s.id] = func(_ string, err error) {
		if err != nil {
			delete(extensionListeners, s.id)
			fail(err)
			return
		}
		s.ready = true
		if s.closed {
			s.stop()
			return
		}
		if ready != nil {
			ready(nil)
		}
	}
	host.Current.Invoke(s.id, "extension/"+name+"/start", string(encoded))
	return s
}

func (s *ExtensionSession) Command(method string, value any, done func(error)) {
	if done == nil {
		done = func(error) {}
	}
	if s == nil || s.closed || !s.ready {
		done(fmt.Errorf("extension is not ready or has been closed"))
		return
	}
	Invoke("extension/"+s.name+"/"+method, map[string]any{"session": s.id, "value": value}, func(_ string, err error) { done(err) })
}

func (s *ExtensionSession) stop() {
	if appReady {
		Invoke("extension/"+s.name+"/stop", map[string]any{"session": s.id}, nil)
	}
}

func (s *ExtensionSession) Close() {
	if s == nil || s.closed {
		return
	}
	s.closed = true
	delete(extensionListeners, s.id)
	if s.ready {
		s.stop()
	}
}

File: packages/native/src/extension.ts
import { app } from "./index.ts";
import * as binding from "./binding.ts";

function validate(name: string, method = "start") {
  if (!/^[a-z][a-z0-9-]{0,63}$/.test(name) || !/^[A-Za-z][A-Za-z0-9_.-]{0,63}$/.test(method))
    throw new TypeError("Invalid native extension name or method");
}
export function invokeExtension<T = unknown>(
  name: string,
  method: string,
  value: unknown = null,
): Promise<T> {
  validate(name, method);
  return binding.invoke(`extension/${name}/${method}`, value);
}
/** One event stream and its requests, owned by the app until closed. */
export class ExtensionSession<T = unknown> {
  readonly ready: Promise<void>;
  readonly #id: number;
  #closed = false;
  readonly #unsubscribe: () => void;
  constructor(
    readonly name: string,
    options: unknown,
    changed: (event: T) => void,
  ) {
    validate(name);
    if (!app.isReady()) throw new Error("Await app.whenReady() before opening an extension");
    const started = binding.startExtension(name, options, (value) => {
      if (!this.#closed) changed(JSON.parse(value));
    });
    this.#id = started.session;
    this.ready = started.ready;
    this.#unsubscribe = app.on("quit", () => {
      this.#closed = true;
    });
    void this.ready.catch(() => {
      this.#closed = true;
      this.#unsubscribe();
    });
  }
  async request<R = unknown>(method: string, value: unknown = null): Promise<R> {
    validate(this.name, method);
    if (method === "start" || method === "stop")
      throw new Error("Use the session constructor and close() for lifecycle");
    await this.ready;
    if (this.#closed) throw new Error("Extension session is closed");
    return invokeExtension(this.name, method, { session: this.#id, value });
  }
  async close(): Promise<void> {
    if (this.#closed) return;
    this.#closed = true;
    this.#unsubscribe();
    try {
      await this.ready;
    } catch {
      return;
    }
    await binding.stopExtension(this.name, this.#id);
  }
}
